examples/Model/User.php

```diff
@@ -12,58 +13,52 @@
 
 namespace Prooph\MicroExample\Model\User;
 
+use Amp\Promise;
+use Amp\Success;
 use InvalidArgumentException;
-use Prooph\Common\Messaging\Message;
+use Phunkie\Types\ImmList;
 use Prooph\MicroExample\Model\Command\ChangeUserName;
 use Prooph\MicroExample\Model\Command\RegisterUser;
-use Prooph\MicroExample\Model\Event\UserNameWasChanged;
-use Prooph\MicroExample\Model\Event\UserWasRegistered;
-use Prooph\MicroExample\Model\Event\UserWasRegisteredWithDuplicateEmail;
+use Prooph\MicroExample\Model\Event\UserNameChanged;
+use Prooph\MicroExample\Model\Event\UserRegistered;
+use Prooph\MicroExample\Model\Event\UserRegisteredWithDuplicateEmail;
 use Prooph\MicroExample\Model\UniqueEmailGuard;
 
 const registerUser = '\Prooph\MicroExample\Model\User\registerUser';
 
-function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): array
+function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): Promise
 {
     if ($guard->isUnique($command->email())) {
-        return [new UserWasRegistered($command->payload())];
+        return new Success(ImmList(new UserRegistered($command->payload())));
     }
 
-    return [new UserWasRegisteredWithDuplicateEmail($command->payload())];
+    return new Success(ImmList(new UserRegisteredWithDuplicateEmail($command->payload())));
 }
```

Comment:
I like that these functions are now async and as such can use async guards. But I'm slightly doubtful about it resolving to an `ImmList` of event. In my project I let them return an [asynchronous Iterator](https://amphp.org/amp/iterators/). What is your thought about that?

Note: Code examples assume that UniqueEmailGuard returns a Promise. Also I highly prefer to use the convention of handling error cases first with early return, leaving the happy scenario as main branch so I negated the condition.

```php
function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): Iterator
{
    return new Producer(function (callable $emit) use ($command, $guard): Generator {
        if (! yield $guard->isUnique($command->email())) {
             yield $emit(new UserRegisteredWithDuplicateEmail($command->payload()));
             return;
        }
        
        yield $emit(new UserRegistered($command->payload()));
    });
}
```

Admittedly while pure code should return Amp\Iterator here, I later decided that I don't like the syntax with `$emit` so I actually let these functions return a Generator which can yield both Promises and Events. Then there is a wrapper function that transforms it into `\Amp\Iterator<EventInterface>`. I consider such syntactic improvement as acceptable over code purity for my project although not sure if it's acceptable for an example like this... But I'll show it to you anyway for comparison:

```php
function registerUser(callable $stateResolver, RegisterUser $command, UniqueEmailGuard $guard): Generator
{
    if (! yield $guard->isUnique($command->email())) {
         yield new UserRegisteredWithDuplicateEmail($command->payload());
         return;
    }

    yield new UserRegistered($command->payload());
}
```

The wrapper function looks something like this:

```php
function transform(Generator $generator): Iterator
{
    return new Producer(function (callable $emit) use ($generator) {
        foreach ($generator as $eventOrPromise) {
            if ($eventOrPromise instanceof Promise) {
                yield $eventOrPromise;
            } else {
                yield $emit($eventOrPromise);
            }
        }
    });
}
```

Reply:
awesome, thanks !!

Reply:
@enumag I ran into issues with this. On `yield $eventOrPromise;` the promises is yielded, but not assigned to the variable in the handler function.
`$state = yield $stateResolver();` results in `$state = null`.

Reply:
Ah yeah... sorry, I copy pasted a wrong function. I'll post the real one tomorrow.

Reply:
thanks

Reply:
maybe can you directly submit a PR to this branch?